### PR TITLE
Proposal: Do expression and async do expression

### PR DESCRIPTION
**Pull request by Jack-Works**
This is an experimental PR to support ECMAScript proposal do expression.

Proposal: [Do expression](https://github.com/tc39/proposal-do-expressions) and [Async do expression](https://github.com/tc39/proposal-async-do-expressions)

Spec: [Do expression](https://tc39.es/proposal-do-expressions/) and [Async do expression](https://tc39.es/proposal-async-do-expressions/)

[Playground]: https://www.staging-typescript.org/play?ts=4.5.0-pr-42437-39

[PR preview Playground][Playground]

## Known problems / TODOs:

- [ ] Completion type analysis should be based on control flow analysis.
- [ ] There're many false positives on the `Not all code paths return a value` check.
- [ ] `break` or `continue` inside the `for` head doesn't lead to a type error ([there is no spec yet, only a memo](https://github.com/tc39/proposal-do-expressions/blob/master/spec.emu#L36) so I don't know how to implement this.)
- [ ] The following code will report type error:

```ts
function f(param: number) {
    const double = do {
        if (param === 0) return null;
        param * param;
    }
    return double
}
```

## Type evaluation rule of the do expression:

I made a change to the checker of typescript. Here are the new evaluation rules:

- Every statement might return `void` or `Type` as a type checking result. If it returns void, it is treated as not contribute to the type of the statement.

#### New Expressions

- **DoExpression**: type of its containing block statement. if there is none, fallback to `voidType`.

#### Statements

`Statement`:
- `BlockStatement`: type is last type of it's inner statement.
- `VariableStatement`: SyntaxError
- `EmptyStatement`: Does not contribute. See https://tc39.es/ecma262/#sec-block-runtime-semantics-evaluation
- `ExpressionStatement`: type of it's containing expression
- `IfStatement`: union type of all branches. if there is no else block, it will treated as `voidType`
- `BreakableStatement`: 
- - `IterationStatement`: SyntaxError
- - `SwitchStatement`: union type of all clauses
- `ContinueStatement`: `never`
- `BreakStatement`: `never`
- `ReturnStatement`: `never`
- `WithStatement`: `any`
- `LabelledStatement`: type of it's inner statement
- `ThrowStatement`: `never`
- `TryStatement`: union type of `try` clause and `catch` clause. `finally` clause is ignored.
- `DebuggerStatement`: Does not contribute. See https://tc39.es/ecma262/#sec-block-runtime-semantics-evaluation

##### A var declaration cannot be used within a do expression unless the target is ESNext

```js
function a() {
    (do {
        var x = 1;
        x;
    });
    x; // 1
}
```

#### Not all code paths return a value

Requires `noImplicitReturns` to be `true`

image

#### Inlay hints

Show all "exit points" of the do expression to make it easier to identify.

image

### CODE REVIEW

**Comment by Jack-Works** [comment]
I need to make a new thing called `checkStatement` that will return a `Type`

**Comment by Jack-Works** [comment]
image

A very naive type-level behavior is done

**Comment by Kingwl** [comment]
Cool! 
Let me (us) know if you want to make a tarball package and a playground.

**Comment by Jack-Works** [comment]
@Kingwl hi most of the semantics is done(except downlevel compiling), can you create a playground for this?

**Comment by Jack-Works** [comment]
oh wait here is CI error

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
@typescript-bot pack this

**Comment by Jack-Works** [comment]
Known problem: transpile will crash if there is `else if` in the do-expr. will fix later.

**Comment by Jack-Works** [comment]
@Kingwl hi I fix lots of bugs plz make a new pack thanks!

**Comment by Kingwl** [comment]
@typescript-bot pack this

**Comment by dragomirtitian** [comment]
Looking through the spec, `() => do { 1 }` should be allowed and is currently an error ([ex](https://www.staging-typescript.org/play?ts=4.2.0-pr-42437-10#code/DYUwLgBAHhC8EAoCUcB8EAmB7CBvCAjBAL4BQQA)). The babel implementation allows it ([ex](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAZjAvDAFASiQPhgExDAbxgCgYYAbAU1gEMkYBGAblJjoGpHiBfIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Cstage-2&prettier=false&targets=&version=7.12.12&externalPlugins=%40babel%2Fplugin-proposal-private-methods%407.12.1%2C%40babel%2Fplugin-proposal-class-properties%407.12.1%2Cbabel-plugin-transform-do-expressions%406.22.0))

The `do` expression is added to `PrimaryExpression` syntax in the [spec](https://bakkot.github.io/do-expressions-v2/#prod-PrimaryExpression), looking at the definition of arrow function we can get to  a Primary expression without adding any tokens:

<details><summary>Full Path</summary>

https://tc39.es/ecma262/#sec-arrow-function-definitions
https://tc39.es/ecma262/#prod-ExpressionBody
https://tc39.es/ecma262/#prod-ConciseBody
https://tc39.es/ecma262/#prod-AssignmentExpression
https://tc39.es/ecma262/#prod-ConditionalExpression
https://tc39.es/ecma262/#prod-ShortCircuitExpression
https://tc39.es/ecma262/#prod-LogicalORExpression
https://tc39.es/ecma262/#prod-CoalesceExpression
https://tc39.es/ecma262/#prod-BitwiseORExpression
https://tc39.es/ecma262/#prod-EqualityExpression
https://tc39.es/ecma262/#prod-RelationalExpression
https://tc39.es/ecma262/#prod-ShiftExpression
https://tc39.es/ecma262/#prod-AdditiveExpression
https://tc39.es/ecma262/#prod-MultiplicativeExpression
https://tc39.es/ecma262/#prod-ExponentiationExpression
https://tc39.es/ecma262/#prod-UnaryExpression
https://tc39.es/ecma262/#prod-UpdateExpression
https://tc39.es/ecma262/#prod-LeftHandSideExpression
https://tc39.es/ecma262/#prod-NewExpression
https://tc39.es/ecma262/#prod-MemberExpression
https://tc39.es/ecma262/#prod-PrimaryExpression
</details>

**Comment by Jack-Works** [comment]
Thanks for your report! I will checkout it later.

**Comment by shadowtime2000** [comment]
Just wondering because of this, is Typescript accepting PRs for ES proposals that haven't been added yet?

**Comment by Kingwl** [comment]
> is Typescript accepting PRs for ES proposals that haven't been added yet?

Nope.

**Comment by Kingwl** [comment]
And the PR will **definitely** not be approved until it's stage 3.

**Comment by Jack-Works** [comment]
No it's must be stage 3 to be accepted. This is just a playground.

Another thing to mention is, do expression proposal is considering to remove the limitation of not able to return continue or break in the do block. So I will not touch those features these days.

I may add async do expressions to this PR later because they are slightly related.

**Comment by Jack-Works** [comment]
@Kingwl hi I've just added support for `async do expression` can you create a new playground?

**Comment by Kingwl** [comment]
Sure.
@typescript-bot pack this.

**Comment by Jack-Works** [comment]
Recently do expression has changed their semantics to allow `break`, `continue`, and `return` across the do expression boundary and it becomes very tricky to compile those expressions. See https://github.com/bakkot/do-expressions-v2/issues/9 for details. I'll consider how to compile this correctly so the emit in those cases are currently invalid.

**Comment by Jack-Works** [comment]
I'd like to ask for some advice from the TypeScript team (I don't know who I should ping, maybe @rbuckton? 👀) about how to implement this.

It is possible to transform do expression into the form of `(expr, expr2, expr3)` if the inner statements met some special requirements. But that need to track and rename variable names in the transformation stage. Consider the following code:

```js
const x = do {
    function y() { return x }
    let x = 1
    y
}
```

Can be transformed into

```js
let _a, _b
const x = ((_a = function y() { return _a }), (_b = 1), _a)
```

But in this transform, I need to track the lexical usage of `x` and `y` and replace them with the temp variable. Should I implement this optimization? The full rule of [the optimized output is documented here](https://blog.jackworks.dev/2021/memo-transfrom-do-expressions/#Part-0-Optimized-output).

If the TypeScript team thought it's worthy to implement this for an inline output, how should I do the usage analysis during the transform stage? If it is not worthy please tell me and I'm not going to code this part.

**Comment by Jack-Works** [comment]
And another question, which way is preferred?

image

(The benefit of `yield` way is I can convert `return-in-do` into a problem `yield-in-do` that already solved.)

**Comment by Jack-Works** [comment]
@Kingwl hi, can you pack this? I fixed some bugs. `break` `continue` and `return` across the boundary are still in progress. Thanks!

**Comment by Kingwl** [comment]
Sure.

@typescript-bot pack this.

**Comment by Jack-Works** [comment]
image

Some progress on control flow changes within do expression. (Note I temporarily disabled the transform of do expression because it does not work well with the control flow transform).

**Comment by Jack-Works** [comment]
image

A bit more progress, supports break/continue in iteration statements (not labeled statement or switch)

**Comment by Jack-Works** [comment]
I have rewritten the transformer to support break/continue/return. The new transformation is new and incomplete but it might handle the most common cases. Note the following items is known not handled correctly:

- `var` hoist across the block
- `return` in the function parameter initializer
- `break`/`continue` in the non-body position of a loop.
- switch

@Kingwl hi, can you pack this? Thanks.

**Comment by Kingwl** [comment]
Sure. You are on 🔥!

@typescript-bot pack this.

**Comment by Jack-Works** [comment]
@Kingwl hi can you pack this thanks!

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1851,6 +1854,9 @@ namespace ts {
                 case SyntaxKind.CaseBlock:
                     return ContainerFlags.IsBlockScopedContainer;
 
+                case SyntaxKind.DoExpression:
+                    return ContainerFlags.IsContainer | ContainerFlags.HasLocals | ContainerFlags.IsBlockScopedContainer;
```
Comment (Jack-Works) [inline src/compiler/binder.ts:1858]: Hi, I don't quite clear about what those flags mean. Is this appropriate? Should I add `IsControlFlowContainer` when it is `async` do expression which does not start a new `return`/`break`/`continue` context. @ahejlsberg can you help me with this problem?

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
Oops, seems I cannot trigger ts bot anymore. 

@orta mighty give you some help.

**Comment by orta** [comment]
@typescript-bot pack this.

**Comment by Jack-Works** [comment]
Updates: 

- Transform: fix `return` inside parameter of a function
- Transform: fix `var` declared variable hoisting
- Transform: support break through `switch` statement
- Transform: fix in class fields
- LSP: support inlay hint for all exit points of the `do` expression

image


@orta Hi can you let the bot pack this?

**Comment by Jack-Works** [comment]
Got some problems with `break` analysis...

**Comment by Jack-Works** [comment]
close for now. I'll re-open and rebase this once the proposal gets advanced.

**Pull request closed without merging**